python/pyiceberg/expressions/__init__.py

```diff
@@ -281,249 +361,313 @@ def __invert__(self) -> BoundIsNull:
         if type(bound_type) in {FloatType, DoubleType}:
             return super().__new__(cls)
         return AlwaysTrue()
 
-    def __invert__(self) -> BoundIsNaN[T]:
+    def __invert__(self) -> BoundIsNaN[L]:
         return BoundIsNaN(self.term)
 
 
-@dataclass(frozen=True)
-class IsNaN(UnaryPredicate[T]):
-    as_bound = BoundIsNaN
-
-    def __invert__(self) -> NotNaN[T]:
+class IsNaN(UnaryPredicate):
+    def __invert__(self) -> NotNaN:
         return NotNaN(self.term)
 
+    @property
+    def as_bound(self) -> Type[BoundIsNaN[L]]:
+        return BoundIsNaN[L]
 
-@dataclass(frozen=True)
-class NotNaN(UnaryPredicate[T]):
-    as_bound = BoundNotNaN
 
-    def __invert__(self) -> IsNaN[T]:
+class NotNaN(UnaryPredicate):
+    def __invert__(self) -> IsNaN:
         return IsNaN(self.term)
 
+    @property
+    def as_bound(self) -> Type[BoundNotNaN[L]]:
+        return BoundNotNaN[L]
 
-@dataclass(frozen=True)
-class SetPredicate(UnboundPredicate[T]):
-    literals: tuple[Literal[T], ...]
 
-    def __invert__(self) -> SetPredicate[T]:
-        """Inverted expression of the SetPredicate"""
-        raise NotImplementedError
+class SetPredicate(Generic[L], UnboundPredicate, ABC):
+    literals: Set[Literal[L]]
 
-    def bind(self, schema: Schema, case_sensitive: bool = True) -> BooleanExpression:
+    def __init__(self, term: Union[str, UnboundTerm], literals: Union[Iterable[L], Iterable[Literal[L]]]):
+        super().__init__(term)
+        self.literals = _to_literal_set(literals)
+
+    def bind(self, schema: Schema, case_sensitive: bool = True) -> BoundSetPredicate[L]:
         bound_term = self.term.bind(schema, case_sensitive)
         return self.as_bound(bound_term, {lit.to(bound_term.ref().field.field_type) for lit in self.literals})
 
+    def __str__(self):
+        # Sort to make it deterministic
+        return f"{str(self.__class__.__name__)}({str(self.term)}, {{{', '.join(sorted([str(literal) for literal in self.literals]))}}})"
+
+    def __repr__(self) -> str:
+        # Sort to make it deterministic
+        return f"{str(self.__class__.__name__)}({repr(self.term)}, {{{', '.join(sorted([repr(literal) for literal in self.literals]))}}})"
+
+    def __eq__(self, other: Any) -> bool:
+        return self.term == other.term and self.literals == other.literals if isinstance(other, SetPredicate) else False
+
+    @property
+    @abstractmethod
+    def as_bound(self) -> Type[BoundSetPredicate[L]]:
+        return BoundSetPredicate[L]
+
+
+class BoundSetPredicate(BoundPredicate[L], ABC):
+    literals: Set[Literal[L]]
 
-@dataclass(frozen=True)
-class BoundSetPredicate(BoundPredicate[T]):
-    literals: set[Literal[T]]
+    def __init__(self, term: BoundTerm[L], literals: Union[Iterable[L], Iterable[Literal[L]]]):
+        # Since we don't know the type of BoundPredicate[L], we have to ignore this one
+        super().__init__(term)  # type: ignore
+        self.literals = _to_literal_set(literals)  # pylint: disable=W0621
```

Comment:
I had expected the `# type: ignore` annotations to be needed in `bind` rather than here because the types already appear to match here.

Reply:
I see your point here, it generates the following error:
```
➜  python git:(fd-remove-dataclass) ✗ pre-commit run --all-files mypy
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

python/pyiceberg/expressions/__init__.py:455: error: Argument 1 to "__init__" of "BoundPredicate" has incompatible type "BoundTerm[str]"; expected "BoundTerm[L]"
python/pyiceberg/expressions/__init__.py:455: error: Argument 1 to "__init__" of "BoundPredicate" has incompatible type "BoundTerm[bool]"; expected "BoundTerm[L]"
python/pyiceberg/expressions/__init__.py:455: error: Argument 1 to "__init__" of "BoundPredicate" has incompatible type "BoundTerm[int]"; expected "BoundTerm[L]"
python/pyiceberg/expressions/__init__.py:455: error: Argument 1 to "__init__" of "BoundPredicate" has incompatible type "BoundTerm[float]"; expected "BoundTerm[L]"
python/pyiceberg/expressions/__init__.py:455: error: Argument 1 to "__init__" of "BoundPredicate" has incompatible type "BoundTerm[bytes]"; expected "BoundTerm[L]"
python/pyiceberg/expressions/__init__.py:455: error: Argument 1 to "__init__" of "BoundPredicate" has incompatible type "BoundTerm[UUID]"; expected "BoundTerm[L]"
python/pyiceberg/expressions/__init__.py:455: error: Argument 1 to "__init__" of "BoundPredicate" has incompatible type "BoundTerm[Decimal]"; expected "BoundTerm[L]"
Found 7 errors in 1 file (checked 83 source files)
```
It does not have a type for `L`, therefore it seems to fail when calling the super constructor. I agree that this is suspicious 🤔